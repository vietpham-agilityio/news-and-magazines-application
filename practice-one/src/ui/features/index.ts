export * from '@/ui/features/TagGroup';
export * from '@/ui/features/SidebarCard';
export * from '@/ui/features/HashtagGroup';

export { default as BadgeGroup } from '@/ui/features/BadgeGroup';
export { default as CommentGroup } from '@/ui/features/CommentGroup';
export { default as CategoryMenu } from '@/ui/features/CategoryMenu';
export { default as AuthorInformation } from '@/ui/features/AuthorInformation';
export { default as CategoryBurgerMenu } from '@/ui/features/CategoryBurgerMenu';

// sections
export { default as PostsGrid } from '@/ui/sections/PostsGrid';
export { default as PostDetail } from '@/ui/sections/PostDetail';
export { default as RightSidebar } from '@/ui/sections/RightSidebar';
export { default as PostsByCategory } from '@/ui/sections/PostsByCategory';
export { default as PostsByAttribute } from '@/ui/sections/PostsByAttribute';