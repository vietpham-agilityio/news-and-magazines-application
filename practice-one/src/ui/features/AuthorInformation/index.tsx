// services
import { getAuthorById, getPostDataById } from '@/services';

// components
import { Avatar, Button, Information } from '@/ui/components';

// icons
import { Plus } from '@/ui/components/Icons';

interface IProps {
  postId: string;
}

const AuthorInformation = async ({ postId }: IProps) => {
  const postResponse = await getPostDataById(postId);
  const { authorId } = postResponse.data.attributes;

  const { data: authorResponse } = await getAuthorById(authorId);
  const { avatarUrl, name } = authorResponse.attributes;

  return (
    <div className="bg-white-90 flex p-[15px] mb-[25px] rounded-xl">
      <Avatar width={87} height={87} url={avatarUrl} alt="Avatar author" />
      <div className="flex-1 ml-2.5 mt-3.5">
        <Information name={name} information="27 posts" />
        <div className="mt-2">
          <Button
            label="Follow"
            leftIcon={<Plus />}
            bgColor="bg-gradient-to-bl from-sunSet-0 to-sunSet-100"
            textColor="text-white-100"
          />
        </div>
      </div>
    </div>
  );
};

export default AuthorInformation;
