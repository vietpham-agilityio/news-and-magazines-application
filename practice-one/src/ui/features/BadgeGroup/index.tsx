// services
import {
  getCategoryById,
  getCategoryByPostId,
  getCommentByPostId,
  getPostDataById,
} from '@/services';

// componens
import { Tag } from '@/ui/components';

// icons
import { ChatBox, Schedule, File } from '@/ui/components/Icons';

// utils
import { formatDate } from '@/utils';

interface IProps {
  postId: string;
}

export default async function BadgeGroup({ postId }: IProps) {
  const postResponse = await getPostDataById(postId);
  const { data: listComments } = await getCommentByPostId(postId);

  const { data: postCategoriesDataResponse } = await getCategoryByPostId({
    postId,
  });
  const categoryId =
    postCategoriesDataResponse[0]?.attributes.categoryId ?? '1';
  const categoryIdToNumber = parseInt(categoryId, 10);
  const { data: categoryDataResponse } =
    await getCategoryById(categoryIdToNumber);
  const { name: categoryType } = categoryDataResponse.attributes;

  const attributes = postResponse?.data?.attributes;
  const publicDate = formatDate(attributes?.publicationDate) ?? '';
  const commentCount = listComments?.length ?? '0';

  return (
    <div className="flex w-full justify-evenly mx-auto mb-12 xl:w-2/4">
      <Tag icon={<Schedule />} content={publicDate} />
      <Tag
        icon={<ChatBox />}
        content={`Comments: ${commentCount}`}
        additionalClass="hidden sm:flex"
      />
      <Tag
        icon={<File />}
        content={`Categories: ${categoryType}`}
        additionalClass="hidden sm:flex"
      />
      <Tag
        icon={<ChatBox />}
        content={`${commentCount}`}
        additionalClass="sm:hidden"
      />
      <Tag
        icon={<File />}
        content={`${categoryType}`}
        additionalClass="sm:hidden"
      />
    </div>
  );
}
