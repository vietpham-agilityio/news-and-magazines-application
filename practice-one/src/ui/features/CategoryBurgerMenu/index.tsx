// services
import { getCategoryData } from '@/services';

// types
import { ICategoryItem, IMenu } from '@/types';

// componens
import { BurgerMenu } from '@/ui/components';

const CategoryBurgerMenu = async () => {
  const { data: categoryDataResponse = [] } = await getCategoryData();

  const listCategories: IMenu[] = [];

  categoryDataResponse?.map((category: ICategoryItem) => {
    const { id } = category;
    const { name } = category.attributes;

    listCategories.push({
      id: id.toString(),
      name: name,
    });
  });

  return <BurgerMenu listCategories={listCategories} />;
};

export default CategoryBurgerMenu;
