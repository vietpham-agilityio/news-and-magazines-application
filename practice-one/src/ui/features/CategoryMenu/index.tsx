// services
import { getCategoryData } from '@/services';
import { ICategoryItem, IMenu } from '@/types';

// componens
import { Menu } from '@/ui/components';

export default async function CategoryMenu() {
  const { data: categoryDataResponse = [] } = await getCategoryData();

  const listCategories: IMenu[] = [];

  categoryDataResponse.map((category: ICategoryItem) => {
    const { id } = category;
    const { name } = category.attributes;

    listCategories.push({
      id: id.toString(),
      name: name,
    });
  });

  return <Menu name="Categories" listMenu={listCategories} />;
}
