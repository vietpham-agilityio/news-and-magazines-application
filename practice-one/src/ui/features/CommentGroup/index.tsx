// components
import { Comment, Typography } from '@/ui/components';

// services
import { getCommentByPostId } from '@/services';

// types
import { FontWeight } from '@/types';

// utils
import { formatDate } from '@/utils';

interface IProps {
  postId: string;
}

const CommentGroup = async ({ postId }: IProps) => {
  const { data: listComments } = await getCommentByPostId(postId);

  if (listComments.length === 0) {
    return (
      <div className="flex justify-center py-10">
        <Typography
          tag="h3"
          weight={FontWeight.Light}
          additionalClasses="text-sm text-black-75 leading-5"
        >
          This article has no comments
        </Typography>
      </div>
    );
  }

  return (
    <div className="flex flex-col gap-9 mb-10 sm:mb-12 lg:mb-15">
      {listComments.map(comment => {
        const { name, content, createdAt } = comment.attributes;
        return (
          <Comment
            key={comment.id}
            userName={name}
            content={content}
            publicDate={formatDate(createdAt)}
          />
        );
      })}
    </div>
  );
};

export default CommentGroup;
