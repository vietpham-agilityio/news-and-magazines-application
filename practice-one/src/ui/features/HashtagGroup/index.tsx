// types
import { IHashtag } from "@/types";
import { LazyImage } from "@/ui/components";

interface IProps {
  listHashtags: IHashtag[];
}

export const HashtagGroup = ({ listHashtags }: IProps) =>
  listHashtags.map((hashtag) => {
    const { id, name, imageUrl } = hashtag;
    return (
      <div key={id} className="mx-3 py-2.5 rounded-xl cursor-pointer">
        <LazyImage
         src={imageUrl}
         alt={name}
         width={170}
         height={48}
        />
      </div>
    );
  });
