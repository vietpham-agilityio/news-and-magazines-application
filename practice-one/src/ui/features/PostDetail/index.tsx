// services
import {
  getCategoryById,
  getCategoryByPostId,
  getCommentByPostId,
  getPostDataById,
} from '@/services';

// utils
import generateRGBDataURL from '@/utils/color';

// componens
import Image from 'next/image';
import { Typography, Tag } from '@/ui/components';

// type
import { FontWeight, Size } from '@/types';

// icons
import { ChatBox, File, Schedule } from '@/ui/components/Icons';
import { formatDate } from '@/utils';

export default async function PostDetail({ postId }: { postId: string }) {
  const postResponse = await getPostDataById(postId);
  const { data: listComments } = await getCommentByPostId(postId);

  const { data: postCategoriesDataResponse } = await getCategoryByPostId({
    postId,
  });
  const categoryId = postCategoriesDataResponse[0]?.attributes.categoryId ?? '1';
  console.log('categoryId', categoryId);
  const categoryIdToNumber = parseInt(categoryId, 10);
  const { data: categoryDataResponse } =
    await getCategoryById(categoryIdToNumber);
  const { name: categoryType } = categoryDataResponse.attributes;

  const attributes = postResponse?.data?.attributes;

  const title = attributes?.title ?? 'Title is not available';
  const content = attributes?.content ?? 'Content is not available';
  const imageUrl = attributes?.imageUrl ?? '';
  const publicDate = formatDate(attributes?.publicationDate) ?? '';
  const commentCount = listComments?.length ?? '0';

  return (
    <section className="flex flex-1">
      <article className="flex-1">
        <div className="mb-7.5 lg:bg-white-90 lg:p-7 xl:mb-32.5 xl:h-151 rounded-xl">
          <Typography
            tag="h1"
            additionalClasses="text-md font-medium sm:text-xxl sm:font-normal mb-7"
          >
            {title}
          </Typography>
          <div className="relative mx-auto w-89.5 h-49.5 sm:w-full sm:h-84 lg:h-95 xl:w-full xl:h-134.5 2xl:h-149.5 2xl:w-268">
            <Image
              fill
              objectFit="cover"
              className="rounded-xl"
              src={imageUrl}
              alt={title}
              placeholder="blur"
              blurDataURL={generateRGBDataURL(64, 64, 64)}
              sizes="(max-width: 603px) 358px,
              (min-width: 604px) and (max-width: 1535px) 100vw,
              (min-width: 1535px) 1072px"
            />
          </div>
        </div>
        <div className="flex w-full justify-evenly mx-auto mb-12 xl:w-2/4">
          <Tag icon={<Schedule />} content={publicDate} />
          <Tag
            icon={<ChatBox />}
            content={`Comments: ${commentCount}`}
            additionalClass="hidden sm:flex"
          />
          <Tag
            icon={<File />}
            content={`Categories: ${categoryType}`}
            additionalClass="hidden sm:flex"
          />
          <Tag
            icon={<ChatBox />}
            content={`${commentCount}`}
            additionalClass="sm:hidden"
          />
          <Tag
            icon={<File />}
            content={`${categoryType}`}
            additionalClass="sm:hidden"
          />
        </div>
        <section>
          {/* content */}
          {content?.split('\n').map((line, index) => (
            <Typography
              key={index}
              tag="h2"
              textSize={Size.MD}
              weight={FontWeight.Normal}
              additionalClasses="text-dark-100"
            >
              {line}
            </Typography>
          ))}
        </section>
      </article>
    </section>
  );
}
