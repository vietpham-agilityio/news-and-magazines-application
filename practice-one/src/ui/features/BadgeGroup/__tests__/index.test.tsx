import { render, waitFor } from '@testing-library/react';
import { expect } from '@jest/globals';

// service
import {
  getPostDataById,
  getCategoryById,
  getCommentByPostId,
  getCategoryByPostId,
} from '@/services';

// component
import { BadgeGroup } from '@/ui/features';

// mock data
import {
  mockCommentData,
  mockPostDetailData,
  mockCategoryResponseData,
  mockPostCategoriesResponse,
} from '@/mocks';

jest.mock('../../../../services', () => ({
  getPostDataById: jest.fn(),
  getCategoryById: jest.fn(),
  getCommentByPostId: jest.fn(),
  getCategoryByPostId: jest.fn(),
}));

describe('BadgeGroup component', () => {
  beforeEach(() => {
    (
      getCommentByPostId as jest.MockedFunction<typeof Object>
    ).mockResolvedValue(mockCommentData);

    (getPostDataById as jest.MockedFunction<typeof Object>).mockResolvedValue(
      mockPostDetailData
    );

    (getCategoryById as jest.MockedFunction<typeof Object>).mockResolvedValue(
        mockCategoryResponseData
    );

    (
      getCategoryByPostId as jest.MockedFunction<typeof Object>
    ).mockResolvedValue(mockPostCategoriesResponse);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('Should render match with snapshot.', async () => {
    const { container } = render(
      await BadgeGroup({
        postId: '1',
      })
    );

    waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});
