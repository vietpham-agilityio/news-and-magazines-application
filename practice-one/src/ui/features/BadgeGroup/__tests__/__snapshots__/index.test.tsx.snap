// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BadgeGroup component Should render match with snapshot. 1`] = `
<div>
  <div
    class="flex w-full justify-evenly mx-auto mb-12 xl:w-2/4"
  >
    <div
      class="flex cursor-pointer undefined"
    >
      <div
        class="mr-2"
      >
        <svg
          fill="none"
          height="15"
          width="13"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4.531 2.5H8.47V1.406c0-.355.273-.656.656-.656.355 0 .656.3.656.656V2.5h1.094c.957 0 1.75.793 1.75 1.75V13c0 .984-.793 1.75-1.75 1.75h-8.75c-.984 0-1.75-.766-1.75-1.75V4.25c0-.957.766-1.75 1.75-1.75h1.094V1.406c0-.355.273-.656.656-.656.355 0 .656.3.656.656V2.5ZM1.688 13c0 .246.19.438.437.438h8.75a.45.45 0 0 0 .438-.438V6H1.687v7Z"
            fill="#3E3232"
            fill-opacity="0.5"
          />
        </svg>
      </div>
      <h2
        aria-level="2"
        class="text-md font-medium text-dark-75 text-xsm leading-4"
        role="heading"
      >
        2024-05-29
      </h2>
    </div>
    <div
      class="flex cursor-pointer hidden sm:flex"
    >
      <div
        class="mr-2"
      >
        <svg
          fill="none"
          height="13"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4.938 5.438c.464 0 .847.41.847.875a.858.858 0 0 1-.848.875.864.864 0 0 1-.875-.875c0-.465.383-.875.875-.875Zm3.062 0c.465 0 .848.41.848.875 0 .437-.41.847-.848.847-.465 0-.875-.383-.875-.848 0-.492.383-.875.875-.875Zm3.063 0c.464 0 .874.41.847.875a.858.858 0 0 1-.848.875.881.881 0 0 1-.874-.875c0-.465.382-.875.874-.875ZM8 .625c3.855 0 6.973 2.598 6.973 5.715S11.855 12 8 12a8.402 8.402 0 0 1-2.543-.383c-.82.574-2.133 1.258-3.8 1.258-.274 0-.52-.137-.602-.41-.082-.246-.055-.52.136-.711 0 0 .848-.93 1.258-1.996C1.547 8.8 1 7.598 1 6.313 1 3.167 4.117.625 8 .625Zm0 10.063c3.117 0 5.633-1.942 5.633-4.376 0-2.406-2.543-4.375-5.66-4.375-3.118 0-5.66 1.97-5.66 4.375 0 1.176.574 2.051 1.066 2.57l.574.602-.3.766a5.116 5.116 0 0 1-.548 1.094 6.104 6.104 0 0 0 1.586-.793l.52-.383.629.191a7.132 7.132 0 0 0 2.16.329Z"
            fill="#3E3232"
            fill-opacity="0.5"
          />
        </svg>
      </div>
      <h2
        aria-level="2"
        class="text-md font-medium text-dark-75 text-xsm leading-4"
        role="heading"
      >
        Comments: 2
      </h2>
    </div>
    <div
      class="flex cursor-pointer hidden sm:flex"
    >
      <div
        class="mr-2"
      >
        <svg
          data-testid="file-icon"
          fill="none"
          height="13"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.723 2.375c.984 0 1.75.793 1.75 1.75v7c0 .984-.793 1.75-1.75 1.75h-10.5c-.985 0-1.75-.766-1.75-1.75v-8.75c0-.957.765-1.75 1.75-1.75h3.226c.465 0 .903.191 1.23.52l1.34 1.23h4.704Zm.437 8.75v-7a.47.47 0 0 0-.437-.438h-5.25l-1.75-1.613a.44.44 0 0 0-.301-.136h-3.2a.45.45 0 0 0-.437.437v8.75c0 .246.192.438.438.438h10.5a.45.45 0 0 0 .437-.438Z"
            fill="#3E3232"
            fill-opacity="0.5"
          />
        </svg>
      </div>
      <h2
        aria-level="2"
        class="text-md font-medium text-dark-75 text-xsm leading-4"
        role="heading"
      >
        Categories: Category
      </h2>
    </div>
    <div
      class="flex cursor-pointer sm:hidden"
    >
      <div
        class="mr-2"
      >
        <svg
          fill="none"
          height="13"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4.938 5.438c.464 0 .847.41.847.875a.858.858 0 0 1-.848.875.864.864 0 0 1-.875-.875c0-.465.383-.875.875-.875Zm3.062 0c.465 0 .848.41.848.875 0 .437-.41.847-.848.847-.465 0-.875-.383-.875-.848 0-.492.383-.875.875-.875Zm3.063 0c.464 0 .874.41.847.875a.858.858 0 0 1-.848.875.881.881 0 0 1-.874-.875c0-.465.382-.875.874-.875ZM8 .625c3.855 0 6.973 2.598 6.973 5.715S11.855 12 8 12a8.402 8.402 0 0 1-2.543-.383c-.82.574-2.133 1.258-3.8 1.258-.274 0-.52-.137-.602-.41-.082-.246-.055-.52.136-.711 0 0 .848-.93 1.258-1.996C1.547 8.8 1 7.598 1 6.313 1 3.167 4.117.625 8 .625Zm0 10.063c3.117 0 5.633-1.942 5.633-4.376 0-2.406-2.543-4.375-5.66-4.375-3.118 0-5.66 1.97-5.66 4.375 0 1.176.574 2.051 1.066 2.57l.574.602-.3.766a5.116 5.116 0 0 1-.548 1.094 6.104 6.104 0 0 0 1.586-.793l.52-.383.629.191a7.132 7.132 0 0 0 2.16.329Z"
            fill="#3E3232"
            fill-opacity="0.5"
          />
        </svg>
      </div>
      <h2
        aria-level="2"
        class="text-md font-medium text-dark-75 text-xsm leading-4"
        role="heading"
      >
        2
      </h2>
    </div>
    <div
      class="flex cursor-pointer sm:hidden"
    >
      <div
        class="mr-2"
      >
        <svg
          data-testid="file-icon"
          fill="none"
          height="13"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.723 2.375c.984 0 1.75.793 1.75 1.75v7c0 .984-.793 1.75-1.75 1.75h-10.5c-.985 0-1.75-.766-1.75-1.75v-8.75c0-.957.765-1.75 1.75-1.75h3.226c.465 0 .903.191 1.23.52l1.34 1.23h4.704Zm.437 8.75v-7a.47.47 0 0 0-.437-.438h-5.25l-1.75-1.613a.44.44 0 0 0-.301-.136h-3.2a.45.45 0 0 0-.437.437v8.75c0 .246.192.438.438.438h10.5a.45.45 0 0 0 .437-.438Z"
            fill="#3E3232"
            fill-opacity="0.5"
          />
        </svg>
      </div>
      <h2
        aria-level="2"
        class="text-md font-medium text-dark-75 text-xsm leading-4"
        role="heading"
      >
        Category
      </h2>
    </div>
  </div>
</div>
`;
